h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -447,18 +449,32 @@ public void modelFailed(Model failedModel) {
 
     /** All visited hyper params permutations, including the current one. */
     private List<int[]> _visitedPermutations = new ArrayList<>();
+    private int _numberOfNonSkippedPermutations = 0;
     private Set<Integer> _visitedPermutationHashes = new LinkedHashSet<>(); // for fast dupe lookup
 
+    ArrayList<Function<Map<String, Object>, Boolean>> _filterFunctions;
```

Comment:
Can we just take a single filter on the input? Does it matter to GS that there are many filters (combined into one)?

Reply:
@michalkurka In current implementation I have specific type of function( stateful one) - so I would say it does matter ( we can't combine them).  Having separate functions also gives a chance to not to apply some functions to the inputs that have already been filtered out by previous functions. But even if it did not matter I would say it violates single responsibility principle - i think it is more clear to provide separate functions.

Reply:
I didn't mean combine on a code level. Combine/compose multiple filters into one to make the GS API accept just a single filter, not a collection.

Reply:
@michalkurka I see. By using `compose()` or `andThen()` we would not be able to differentiate between functions -  we would only have one `apply()` method which would call chain of functions inside. But I need to behave differently based on the type of the function -  so I think composing is not possible in the given context. In case of Materialised grid I had a way to avoid stateful functions, in case of lazy evaluation we need to remember state somehow -> it led to an introduction of the class that extends function and keeps state.